### PR TITLE
Add response timeout to http client

### PR DESCRIPTION
**Pull request by rpanchyk**

### CODE REVIEW

File: src/test/java/org/prebid/server/vertx/http/BasicHttpClientTest.java [hunk 1]
```diff
@@ -114,13 +132,111 @@ public void requestShouldFailsIfHttpResponseFails() {
...
+        assertThat(future.cause())
+                .isInstanceOf(TimeoutException.class)
+                .hasMessage("Timed out while waiting for response");
+    }
+
+    /**
+     * The server returns entire response or body with delay.
+     */
+    private static void startServer(TestContext context, int port, long entireResponseDelay, long bodyResponseDelay) {
+        final Async async = context.async();
```
Comment (schernysh) [inline src/test/java/org/prebid/server/vertx/http/BasicHttpClientTest.java]: I would say this is misuse of `TestContext`. `CountDownLatch` would be much more suitable here.

File: src/main/java/org/prebid/server/vertx/http/BasicHttpClient.java [hunk 2]
```diff
@@ -24,32 +33,92 @@ public BasicHttpClient(io.vertx.core.http.HttpClient httpClient) {
...
-                .exceptionHandler(future::fail);
+                .bodyHandler(buffer -> successResponse(buffer.toString(), response, future, timerId))
+                .exceptionHandler(exception -> failResponse(exception, future, timerId));
+    }
+
+    private void successResponse(String body, io.vertx.core.http.HttpClientResponse response,
+                                 Future<HttpClientResponse> future, long timerId) {
+        vertx.cancelTimer(timerId);
+
+        future.complete(HttpClientResponse.of(response.statusCode(), response.headers(), body));
```
[the remark this answers is not in the document]
Comment (schernysh) [reply src/main/java/org/prebid/server/vertx/http/BasicHttpClient.java]: Is it possible that success handler will be called after timer completed this future due to timeout? Even if it is unlikely it wouldn't hurt to protect ourselves against it. However we can leave it as is and see if it appears in the logs - it is pretty harmless.

[the remark this answers is not in the document]
Comment (rpanchyk) [reply src/main/java/org/prebid/server/vertx/http/BasicHttpClient.java]: Ok, go through 2nd way, just leave it as is so far.

File: src/main/java/org/prebid/server/vertx/http/BasicHttpClient.java [hunk 3]
```diff
@@ -24,32 +33,92 @@ public BasicHttpClient(io.vertx.core.http.HttpClient httpClient) {
...
-                                       Future<HttpClientResponse> future) {
+    /**
+     * Returns Vert.x Timer ID which fails the given {@link Future} in case of response obtaining is timed out.
+     */
+    private long setResponseTimeout(long timeoutMs, HttpClientRequest httpClientRequest,
+                                    Future<HttpClientResponse> future) {
+        final TimerId timerId = new TimerId();
+
+        // Start timer after the connection is established to be in-sync with request timeout.
+        httpClientRequest.connectionHandler(connection -> timerId.setValue(
```
[the remark this answers is not in the document]
Comment (rpanchyk) [reply src/main/java/org/prebid/server/vertx/http/BasicHttpClient.java]: Full agree, great catch!

**Pull request merged**